include/deal.II/lac/precondition.h

```diff
@@ -1894,51 +1929,69 @@ namespace internal
-      VectorUpdatesRange<Number>(upd, src.size());
+                                solution_old.begin(),
+                                temp_vector1.begin(),
+                                solution.begin());
+      VectorUpdatesRange<Number>(upd, rhs.size());
+
+      // swap vectors x^{n+1}->x^{n}, given the updates in the function above
+      if (iteration_index == 1)
+        {
+          solution.swap(temp_vector1);
+          solution_old.swap(temp_vector1);
+        }
+      else if (iteration_index > 1)
+        solution.swap(solution_old);
     }
 
     // selection for diagonal matrix around parallel deal.II vector
     template <typename Number, typename MemorySpace>
     inline void
     vector_updates(
-      const LinearAlgebra::distributed::Vector<Number, MemorySpace> &src,
+      const LinearAlgebra::distributed::Vector<Number, MemorySpace> &rhs,
       const DiagonalMatrix<
         LinearAlgebra::distributed::Vector<Number, MemorySpace>> &jacobi,
-      const bool                                                  start_zero,
-      const double                                                factor1,
-      const double                                                factor2,
-      LinearAlgebra::distributed::Vector<Number, MemorySpace> &   update1,
-      LinearAlgebra::distributed::Vector<Number, MemorySpace> &   update2,
+      const unsigned int                                       iteration_index,
+      const double                                             factor1,
+      const double                                             factor2,
+      LinearAlgebra::distributed::Vector<Number, MemorySpace> &solution_old,
+      LinearAlgebra::distributed::Vector<Number, MemorySpace> &temp_vector1,
       LinearAlgebra::distributed::Vector<Number, MemorySpace> &,
-      LinearAlgebra::distributed::Vector<Number, MemorySpace> &dst)
+      LinearAlgebra::distributed::Vector<Number, MemorySpace> &solution)
     {
-      VectorUpdater<Number> upd(src.begin(),
+      VectorUpdater<Number> upd(rhs.begin(),
                                 jacobi.get_vector().begin(),
-                                start_zero,
+                                iteration_index,
                                 factor1,
                                 factor2,
-                                update1.begin(),
-                                update2.begin(),
-                                dst.begin());
-      VectorUpdatesRange<Number>(upd, src.local_size());
+                                solution_old.begin(),
+                                temp_vector1.begin(),
+                                solution.begin());
+      VectorUpdatesRange<Number>(upd, rhs.local_size());
+
+      // swap vectors x^{n+1}->x^{n}, given the updates in the function above
+      if (iteration_index == 1)
+        {
+          solution.swap(temp_vector1);
+          solution_old.swap(temp_vector1);
+        }
+      else if (iteration_index > 1)
```

Comment:
iteration_index is always positive, which makes this logic somewhat weird (as this check is unnecessary). Is there a reason you don't start with ``iteration=0``? (here and above of course)

Reply:
wait: that is not true, you also call it with 0. Does it make sense to make an empty
```
if (iteration_index==0)
{
// nothing to do here
}
```
?

Reply:
Yes, it would make sense to add this part to make it similar to the other if statements cases.